feat: support byte ranges

rvagg opened the pull request:
Initial exploration work; it seems like this might not be too tricky after all.

rvagg commented:
The failures in here are:

```
    --- FAIL: TestHttpFetch/graphsync_nested_large_sharded_file,_with_path,_dag-scope_entity (0.23s)
    --- FAIL: TestHttpFetch/bitswap_nested_large_sharded_file,_with_path,_dag-scope_entity (0.21s)
    --- FAIL: TestHttpFetch/http_nested_large_sharded_file,_with_path,_dag-scope_entity (0.13s)
```

The overlap they have with this is that they use a `unixfs-preload` with a matcher, and our verifier now does something with matches (read all bytes from a `LargeBytesNode`). Since `dag-scope=entity` is doing this already via `unixfs-preload`, so these failures should just be that we're seeing the same blocks twice because we read+discard them in two places. That'll need some unifying (or maybe just removing `unixfs-preload` when we're doing `entity` + byte ranges?).

hannahhoward commented:
@rvagg completely forgot to ever push this -- I started a long while ago. Yes, I think what we need to do is turn off pre-load on range requests. The only question then is what to do about UnixFSDirectories. I don' t know how that is supposed to behave with `dag-scope=entity` & `entity-bytes`. https://github.com/filecoin-project/lassie/pull/307

hannahhoward commented:
That code is super old you don't need to use

hannahhoward commented:
Looking at the IPIP-402 proposal, this is gonna be a bit tricky, cause when the node is a directory, `entity-bytes` is supposed to have no effect, i.e. send the whole thing. So then we need preload for directories, non-preload for files. Eek.

hannahhoward commented:
One suggestion: maybe preload should only do directories. That's all you actually need it for.

Then if there is no byte range selector, we simply match the whole file.

That would be a breaking change to preload, so it'd need to get to boost for the purposes of GraphSync.

rvagg commented:
This now depends on https://github.com/ipfs/go-unixfsnode/pull/52 and also https://github.com/ipld/go-ipld-prime/pull/529; those are what are being pointed to in the current go.mod.

The test fixtures now include gobs of ranges reading from a file, including lots of boundary cases.

Things still to do:

* Negative ranges — I could defer this feature to a separate batch of work, it's nontrivial
* Handle the case where the entity at the path terminus is _not_ a file and can't be read as a byte range — this is also potentially very nontrivial since selector's don't really have an "or" for this case and the slice matcher fails hard when it's not interpreting bytes.

rvagg commented:
Added support for the "or" part of 402 - when specifying entity-bytes and the terminus _isn't_ a file, it should behave as if you didn't specify entity-bytes.  This required removing the hard error in the slice matcher; see https://github.com/ipld/go-ipld-prime/pull/529.

rvagg commented:
Read aside from needing these merged:

* [ ] https://github.com/ipld/go-ipld-prime/pull/529
* [ ] https://github.com/ipfs/go-unixfsnode/pull/52

The missing piece here is negative byte ranges. Those will be rejected as unsupported at the moment. Further work in a later PR.

Important to note that only Bitswap requests (and some HTTP!) are going to work with this at the moment:

* HTTP is going to need eipfs to support this, we'll reject their CAR streams in the verifier without it.
* Graphsync needs the go-ipld-prime and go-unixfsnode changes, or there's a disagreement between the peers about what the blocks to exchange are.

BUT https://github.com/ipld/frisbii/pull/9 uses this, and I have a Frisbii node online with that, serving wikipedia and also birb.mp4 (bafybeic56z3yccnla3cutmvqsn5zy3g24muupcsjtoyp3pu5pm5amurjx4) so you can play with that at least.

rvagg commented on pkg/internal/testdata/unixfs_20m_variety.car:
Yeah, I was thinking this was going to be handled by the go pkg install tooling and ignored since it's in `testdata`, but I confirmed that it's in my ~/go/pkg when used as a dependency, which is .. unfortunate. I don't really want this to be a burden for downloaders.

We could git submodule it and put it somewhere else (I have more to make for this). Or do you have a better suggestion?

willscott replied:
since you have code for regenerating it, can you put the fixture itself in gitignore?

rvagg replied:
It's a bit bespoke; I _might_ be able to regenerate it exactly as is but it was made to be fixed so I could inspect it and manually make fixtures that point to it.

I've been thinking that maybe putting this into ipld/ipld along with other fixtures might be the way to deal with this. Then either export a golang package from there or submodule it like we do with go-ipld-prime. We can have a section about paths traversal and unixfs and the IPLD specifics of the trustless gateway spec as it relates to these matters. I was toying with the idea of an ipld/ipld-fixtures repo, but that's really what ipld/ipld was meant to be so maybe try and use that. I'll experiment a bit and see.

hannahhoward replied:
Like I think we can safely assume there will be more test cases we haven't thought of yet coming.

rvagg replied:
Yes, and I wanted to build a corpus of actual data and descriptive tests that can be run against them—so they can be run any place, including Frisbii and others. We currently have a lot of random-generated on-the-fly data (which I keep on having to squash flakies on because of edge cases) but the intent here is to have something fixed that can be reasoned about and inspected separately to the code that generated it.

hannahhoward commented on pkg/verifiedcar/verifiedcar.go:
```diff
@@ -55,8 +55,6 @@ type Config struct {
 	MaxBlocks          uint64         // set a budget for the traversal
```
eek so, I just realized that we're gonna need to figure out what to do with .storage, as they are not yet supporting ranges. I can pop over and try to add it but... in the meantime, maybe we need to figure out some content negotiation in the http retriever.

hannahhoward replied:
otherwise we're looking at 100% failure here.

rvagg replied:
Not 100%. Bitswap will work, Frisbii will work, but yeah the rest will bork if we start pushing them through.

I had some thoughts about negotiation over in 412: https://github.com/ipfs/specs/pull/412#discussion_r1247310092

If we were to request a byte range but then be told the server doesn't support it, at least we could switch modes to parsing their response. What we do on the way out of Lassie is another matter—do we give the user more than they asked for and just give our own negotiation signal that we don't support range cause we couldn't find anyone else that does? Either way this is really hard.

hannahhoward replied:
technically if we knew that we were getting a full response, we could modify the verifier to just skip blocks till you got to the expected blocks in a range request (since technically a full request contains all the blocks for a range request in the correct order, just with other blocks in between). In fact, turning this on when entity-bytes is present by default might be the hack needed to get this over the line.

hannahhoward commented on pkg/verifiedcar/verifiedcar.go:
```diff
@@ -142,62 +136,107 @@ func (cfg Config) VerifyBlockStream(ctx context.Context, cbr BlockReader, lsys l
 	return bt.blocks, bt.bytes, nil
 }
 
-func (cfg *Config) nextBlockReadOpener(ctx context.Context, cr *carReader, bt *writeTracker, lsys linking.LinkSystem) linking.BlockReadOpener {
```
a bunch of this code should be easier to review with #338 merged?

hannahhoward reviewed (commented):
So I don't see any big red flags, but I think we need to talk through the implications of deploying this (esp in Saturn) before we merge.

rvagg commented:
Rebased to main and I extracted the large fixture file to https://github.com/ipld/ipld/pull/290 (open for discussion!). I removed and squashed it out of the commits on this branch so it shouldn't weigh it down:

```
$ git rev-list --disk-usage --objects HEAD...main
391083
```

rvagg commented:
Ready for deeper review, current blockers for merge are these:

* [x] https://github.com/ipld/go-ipld-prime/pull/529
* [ ] https://github.com/ipfs/go-unixfsnode/pull/52 - ready for review
* [ ] https://github.com/ipld/ipld/pull/290 - ready for review (or for someone to vehemently disagree with this approach and ask for something different)
